sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -394,83 +410,349 @@ struct codec_info codec_info_list[] = {
+	"SDW%d-Capture",
+};
 
-#if IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
-	{
-		.name = "iDisp1",
-		.id = 5,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
-	},
-	{
-		.name = "iDisp2",
-		.id = 6,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
-	},
-	{
-		.name = "iDisp3",
-		.id = 7,
-		.init = hdmi_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
-	},
+static inline int find_codec_info(unsigned int part_id)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(codec_info_list); i++)
+		if (part_id == codec_info_list[i].id)
+			break;
+
+	if (i == ARRAY_SIZE(codec_info_list))
+		return -EINVAL;
+
+	return i;
+}
+
+static int get_sdw_dai_link_num(struct snd_soc_acpi_mach_params *mach_params)
+{
+	const struct snd_soc_acpi_link_adr *link;
+	int num = 0;
+	int id;
+
+	link = mach_params->links;
+	if (!link)
+		return -EINVAL;
+
+	while (link->num_adr) {
+		int part_id;
+
+		part_id = (link->adr[0] >> 8) & GENMASK(15, 0);
+		id = find_codec_info(part_id);
+		if (id < 0)
+			return 0;
+
+		/* support playback */
+		if (codec_info_list[id].direction[0])
+			num++;
+
+		/* support capture */
+		if (codec_info_list[id].direction[1])
+			num++;
+
+		link++;
+	}
+
+	return num;
+}
+
+static void init_dai_link(struct snd_soc_dai_link *dai_links,
+			  int id, char *name,
+			  struct snd_soc_dai_link_component *cpus,
+			  char *cpu_name,
+			  struct snd_soc_dai_link_component *codecs,
+			  int codecs_num,
+			  int playback,
+			  int capture)
+{
+	dai_links[id].id = id;
+	dai_links[id].name = name;
+	dai_links[id].platforms = platform_component;
+	dai_links[id].num_platforms = ARRAY_SIZE(platform_component);
+	dai_links[id].nonatomic = true;
+	dai_links[id].no_pcm = 1;
+	dai_links[id].cpus = &cpus[id];
+	dai_links[id].num_cpus = 1;
+	dai_links[id].cpus->dai_name = cpu_name;
+	dai_links[id].codecs = codecs;
+	dai_links[id].num_codecs = codecs_num;
+	dai_links[id].dpcm_playback = playback;
+	dai_links[id].dpcm_capture = capture;
+}
+
+static bool check_unique_id(const struct snd_soc_acpi_link_adr *link,
+			    unsigned int sdw_version,
+			    unsigned int unique_id,
+			    unsigned int mfg_id,
+			    unsigned int part_id,
+			    unsigned int class_id
+			   )
+{
+	unsigned int sdw1_version, unique1_id, mfg1_id, part1_id, class1_id;
+	int i;
+
+	for (i = 0; i < link->num_adr; i++) {
+		sdw1_version = (link->adr[i] >> 44) & GENMASK(4, 0);
+		unique1_id = (link->adr[i] >> 40) & GENMASK(3, 0);
+		mfg1_id = (link->adr[i] >> 24) & GENMASK(15, 0);
+		part1_id = (link->adr[i] >> 8) & GENMASK(15, 0);
+		class1_id = link->adr[i] & GENMASK(7, 0);
+
+		/* find the same codec with different unique_id */
+		if (sdw_version == sdw1_version &&
+		    mfg_id == mfg1_id &&
+		    part_id == part1_id &&
+		    class_id == class1_id &&
+		    unique_id != unique1_id)
```

Comment:
you should only use the unique ID if there are more than one identical devices on the link.

Reply:
done